components/eamxx/src/diagnostics/tests/field_at_height_tests.cpp

```diff
@@ -92,135 +123,200 @@ TEST_CASE("field_at_height")
+  REQUIRE(! views_are_equal(z_int,geo_int));
+  REQUIRE(! views_are_equal(z_mid,geo_mid));
+
+  // Make sure that an unsupported reference height throws an error.
+  print(" -> Testing throws error with unsupported reference height...\n");
+  {
+    REQUIRE_THROWS(run_diag (s_mid,geo_mid,"1m","foobar"));
+  }
+  print(" -> Testing throws error with unsupported reference height... OK\n");
 
   // Run many times
-  Real z_tgt,lev_tgt;
+  int z_tgt,lev_tgt;
   std::string loc;
-  for (int irun=0; irun<nruns; ++irun) {
-
-    // Randomize fields
-    for (auto f : {s_mid, s_int, v_mid, v_int}) {
-      // We know f is not a subfield, so we can use this power-user method
-      auto data = f.get_internal_view_data<Real,Host>();
-      const auto& size = f.get_header().get_identifier().get_layout().size();
-      for (int i=0; i<size; ++i) {
-        data[i] = pdf_fields(engine);
-      }
-      f.sync_to_dev();
-    }
+  for (std::string surf_ref : {"sealevel","surface"}) {
+    printf(" -> Testing for a reference height above %s...\n",surf_ref.c_str());
+    const auto mid_src = surf_ref == "sealevel" ? z_mid : geo_mid;
+    const auto int_src = surf_ref == "sealevel" ? z_int : geo_int;
+    for (int irun=0; irun<nruns; ++irun) {
 
-    z_tgt = pdf_levs(engine);
-    loc = std::to_string(z_tgt) + "m";
-    // z[ilev] = nlevs-ilev, so the tgt slice is nlevs-z_tgt
-    lev_tgt = nlevs-z_tgt;
-
-    print(" -> Testing with z_tgt coinciding with a z level\n");
-    {
-      print("    -> scalar midpoint field...............\n");
-      auto d = run_diag (s_mid,z_mid,loc);
-      auto tgt = s_mid.subfield(1,static_cast<int>(lev_tgt));
-      REQUIRE (views_are_equal(d,tgt,&comm));
-      print("    -> scalar midpoint field............... OK!\n");
-    }
-    {
-      print("    -> scalar interface field...............\n");
-      auto d = run_diag (s_int,z_int,loc);
-      // z_mid = nlevs+1-ilev, so the tgt slice is nlevs+1-z_tgt
-      auto tgt = s_int.subfield(1,static_cast<int>(lev_tgt));
-      REQUIRE (views_are_equal(d,tgt,&comm));
-      print("    -> scalar interface field............... OK!\n");
-    }
-    {
-      print("    -> vector midpoint field...............\n");
-      auto d = run_diag (v_mid,z_mid,loc);
-      // We can't subview over 3rd index and keep layout right,
-      // so do all cols separately
-      for (int i=0; i<ncols; ++i) {
-        auto fi = v_mid.subfield(0,i);
-        auto di = d.subfield(0,i);
-        auto tgt = fi.subfield(1,static_cast<int>(lev_tgt));
-        REQUIRE (views_are_equal(di,tgt,&comm));
+      // Randomize fields using f_z_src function defined above:
+      auto slope = pdf_m(engine); 
+      auto inter = pdf_y0(engine);
+      f_z_src(inter, slope, mid_src, s_mid);
+      f_z_src(inter, slope, mid_src, v_mid);
+      f_z_src(inter, slope, int_src, s_int);
+      f_z_src(inter, slope, int_src, v_int);
+
+      // Set target z-slice for testing to a random value.
```

Comment:
This only tests the case where `z_tgt` matches exeactly one of the levels though. Since you compute src and tgt via analytic function, you can probably switch `pdf_levs` to be a RPDF, and test also intermediate cases.

Reply:
oh yes, good point.  Thanks.  I'll make the change.

Reply:
@bartgol , actually it still needs to be an INT not a REAL because FieldAtHeight assumes an integer number of meters above the reference.  I.e. doesn't support decimals in the height.  So instead I will bump up the `z_top` and made the PDF go from 1 to z_top.

Reply:
Ok, but we need to test the case where `z_tgt` does not coincide with any of the values in `z_mid`/`z_int`. Perhaps you could make them have a stride >=2, so that it is possible that the IPDF generates an integer that is not a value of z?